### PR TITLE
Support paste.gg, hastebin, and mclo.gs

### LINKED ISSUE #342: [Feature] [Pastebin Service] Support mclo.gs as a pastebin service
### Role

Play and Maintain Modpacks

### Suggestion

use mclo.gs as an pastebin service for logs

### Benefit

easier way for people to read logs

### This suggestion is unique

- [x] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

https://mclo.gs/

Legit, makes uploading and reading logs easier as they automatically censor information such as IP addresses and Computer Usernames, 

here is an example from my vanilla+ modpack
https://mclo.gs/sShksYb

I am mainly suggesting this as people are hesitant to send me, and other admins their logs due to the fact that the logs, more often than not, contain private information such as their name, and using mclo.gs as an option allows people to feel safe sending me and other server admins/modpack maintainers their logs

### LINKED ISSUE #419: SUPPORT HST.SH
### Role

Log uploading

### Suggestion

Support hst.sh as a log uploader

### Benefit

Alternative log uploading services 

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### LINKED ISSUE #441: Support websites that don't use `https://0x0.st`'s format
### Role

I want to be able to access easily for more crashlog upload webistess

### Suggestion

Support websites that don't use `https://0x0.st`'s format

### Benefit

Support more crashlog upload websites (eg. https://crashy.net)

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### PR DESCRIPTION
**Pull request by LennyMcLennington**
Closes #342, closes #419, closes #441
also @Scrumplex how do i do the copyright headers properly

### CODE REVIEW

**Comment by Zetvue** [comment]
Some checks haven’t completed yet
6 successful and 4 in progress checks
@github-actions
Build Application / Build Debug / build (ubuntu-20.04) (pull_request) Successful in 2m
[Details](https://github.com/PolyMC/PolyMC/runs/6396394217?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (ubuntu-20.04) (push) Successful in 2m
[Details](https://github.com/PolyMC/PolyMC/runs/6396384937?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (ubuntu-20.04, true) (pull_request) Successful in 4m
[Details](https://github.com/PolyMC/PolyMC/runs/6396394286?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (ubuntu-20.04, true) (push) Successful in 4m
[Details](https://github.com/PolyMC/PolyMC/runs/6396384993?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (windows-2022, Windows-i686, mingw32) (pull_request) In progress — This check has started...
[Details](https://github.com/PolyMC/PolyMC/runs/6396394396?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (windows-2022, Windows-i686, mingw32) (push) In progress — This check has started...
[Details](https://github.com/PolyMC/PolyMC/runs/6396385047?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (windows-2022, Windows-x86_64, mingw64) (pull_request) In progress — This check has started...
[Details](https://github.com/PolyMC/PolyMC/runs/6396394472?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (windows-2022, Windows-x86_64, mingw64) (push) In progress — This check has started...
[Details](https://github.com/PolyMC/PolyMC/runs/6396385099?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (macos-11, 10.13) (pull_request) Successful in 4m
[Details](https://github.com/PolyMC/PolyMC/runs/6396394531?check_suite_focus=true)
@github-actions
Build Application / Build Debug / build (macos-11, 10.13) (push) Successful in 7m
[Details](https://github.com/PolyMC/PolyMC/runs/6396385157?check_suite_focus=true)
This pull request is still a work in progress
Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can mark a draft pull request as ready for review.

**Comment by Zetvue** [comment]
works!

**Comment by Scrumplex** [comment]
> how do i do the copyright headers properly

Just do what you did in APIPage.cpp if there already is a header. If not copy it over from another file

File: launcher/Application.cpp
Comment (Scrumplex) [inline launcher/Application.cpp]: We could switch users who have not changed the URL to mclo.gs as well

File: launcher/ui/pages/global/APIPage.ui
Comment (Scrumplex) [inline launcher/ui/pages/global/APIPage.ui:7]: Can you open the .ui file once with Qt Designer and close it again, so it can figure out formatting and ordering and stuff?

File: launcher/Application.cpp
Comment (LennyMcLennington) [reply launcher/Application.cpp]: ?

**Comment by DioEgizio** [comment]
Also this closes #342 and #419

**Comment by txtsd** [comment]
@LennyMcLennington `hastebin.com` has been acquired by `toptal.com` and their website and endpoints will be changing:

> ## Deprecation notice in the API endpoint
> 
> This website is changing from https://hastebin.com to https://www.toptal.com/developers/hastebin and the API endpoints will change:
> 
> ```
> $ curl -s -X POST https://www.toptal.com/developers/hastebin/documents -d "Hello World!"
> {"key":"aeiou"}
> $ curl https://www.toptal.com/developers/hastebin/raw/aeiou
> Hello World!
> ```

`https://hst.sh/` might be a good alternative.

**Comment by DioEgizio** [comment]
> https://hst.sh/ might be a good alternative.

Why not just using https://www.toptal.com/developers/hastebin as endpoint?

**Comment by txtsd** [comment]
It's a talent search firm. No idea what they'll be doing with paste data.

File: launcher/ui/pages/global/APIPage.ui
Comment (LennyMcLennington) [reply launcher/ui/pages/global/APIPage.ui:7]: What? I used designer to design this ui file.

**Comment by DioEgizio** [comment]
immagine

maybe change this to: 
mclo.gs
0x0.st
paste.gg
hastebin

it'd look more clear

**Comment by LennyMcLennington** [comment]
> immagine
> 
> maybe change this to: mclo.gs 0x0.st paste.gg hastebin
> 
> it'd look more clear

Well the official names as far as I can see are: mclogs, 0x0, paste, haste
So maybe that should be what is written. The point of that is to specify the type of pastebin, not the domain name of a specific instance.

**Comment by DioEgizio** [comment]
Right, but this could confuse some people

**Comment by LennyMcLennington** [comment]
> Right, but this could confuse some people

Okay I will change it.

File: launcher/Application.cpp
Comment (LennyMcLennington) [reply launcher/Application.cpp]: I don't know if it's good to switch people who selected 0x0.st because they _want_ to use 0x0.st over to mclo.gs completely automatically.

But then again, I don't know how many people actually care because I certainly don't, because I am one of the people who just uses 0x0.st because it was the default and it was the simplest protocol to support, not because I actually wanted to use specifically 0x0.st.

File: launcher/net/PasteUpload.cpp
Comment (txtsd) [inline launcher/net/PasteUpload.cpp:49]: Can you make the text say the domain `hst.sh` like the others?

**Review by txtsd — CHANGES_REQUESTED** [review]

File: launcher/net/PasteUpload.cpp
Comment (LennyMcLennington) [reply launcher/net/PasteUpload.cpp:49]: > Can you make the text say the domain `hst.sh` like the others?

The point of that is to specify the type of pastebin, not the domain name of a specific instance. The reason the other ones are domain names is because the names aren't really unique enough and the service itself is better known by the domain name of the official instance. (like, I don't know that someone would know what you're talking about if you say "paste" instead of "paste.gg" because "paste" is the most generic name ever for a pastebin)

It used to say "mclogs", "0x0", "paste (paste.gg)", and "hastebin" but DioEgizio said that it would cause confusion to not have the domain names for those first three and that it should be changed to just use the domain names of the official instance but hastebin didn't have that problem. If anything it would've been changed to hastebin.com, because that was the domain name of the official instance. hst.sh is just the default haste instance that I decided to use instead of hastebin.com because of Toptal's acquisition of hastebin.com.

Comment (txtsd) [reply launcher/net/PasteUpload.cpp:49]: Fair enough.

**Review by txtsd — APPROVED** [review]
My default selection of 0x0.st persisted. Uploads to all paste sites were perfect.

**Comment by DioEgizio** [comment]
For the hastebin stuff; I just noticed hst.sh might use an older version of hastebin, because some stuff is more outdated and it's much worse on mobile.
Also I can't find any information about hst.sh on the internet

File: launcher/Application.cpp
Comment (DioEgizio) [reply launcher/Application.cpp]: people who actually want 0x0.st could still change it manually back

**Comment by DioEgizio** [comment]
immagine
is this normal? (trying to upload setting the hastebin url back to https://hastebin.com or to https://www.toptal.com/developers/hastebin )

**Comment by DioEgizio** [comment]
also this closes https://github.com/PolyMC/PolyMC/issues/441 probably

**Comment by RaptaG** [comment]
~~What about [crashy.net](https://crashy.net)?~~

**Comment by DioEgizio** [comment]
That site has no api and it's not good at all, which is for example why on quilt they request you to not using that site

File: launcher/net/PasteUpload.cpp
Comment (RaptaG) [inline launcher/net/PasteUpload.cpp]: (Suggestion) Maybe rename to "HasteBin", "Paste (paste.gg)" and "MCLogs" (appearance purposes)

Comment (LennyMcLennington) [reply launcher/net/PasteUpload.cpp]: Hastebin isn't called HasteBin it's called Hastebin. Also you're reviewing an older commit, those were changed.

File: launcher/Application.cpp
Comment (LennyMcLennington) [reply launcher/Application.cpp]: > people who actually want 0x0.st could still change it manually back

Yeah, if I will do that I need to add some sort of indication to people that the default paste service was changed and an option to change it back to 0x0.st

**Comment by DioEgizio** [comment]
Why draft?

**Comment by LennyMcLennington** [comment]
> Why draft?

found a bug so i put it to wip until i had time to fix it, but i fixed it now

File: launcher/Application.cpp
Comment (LennyMcLennington) [reply launcher/Application.cpp]: Done

**Comment by LennyMcLennington** [comment]
To test the confirmation dialog thing for people who already used 0x0.st you can put `PastebinURL=https://0x0.st` in polymc.cfg before starting PolyMC. You'll get a dialog on startup that tells you the default changed and asks you to choose whether to switch to the new default or keep your previous settings.

File: launcher/Application.cpp
Comment (Scrumplex) [inline launcher/Application.cpp]: nit: IMO this is easier to read

```suggestion
            bool userHadDefaultPastebin = pastebinURL == "https://0x0.st";
            if (!pastebinURL.isEmpty() && !userHadDefaultPastebin)
```

File: launcher/Application.cpp
Comment (Scrumplex) [inline launcher/Application.cpp:928]: Will this ever happen? The constructor of `Application` will already migrate existing settings.

File: launcher/net/PasteUpload.cpp
Comment (Scrumplex) [inline launcher/net/PasteUpload.cpp:131]: What Qt version is this targeting? Most of the codebase uses the `SIGNAL` and `SLOTG` macros

**Review by Scrumplex — CHANGES_REQUESTED** [review]

File: launcher/net/PasteUpload.cpp
Comment (flowln) [reply launcher/net/PasteUpload.cpp:131]: As far as I know, those macros connect the signals and slots by name, at runtime, not at compile-time. So even though we use those macros throughout the codebase, it's safer and more performant to define it like this, whenever we can.

File: launcher/Application.cpp
Comment (LennyMcLennington) [reply launcher/Application.cpp:928]: > Will this ever happen? The constructor of `Application` will already migrate existing settings.

See for yourself. The auto migration migrates people only if they had not used the default paste URL previously. This is triggered if the user had the default paste URL, and then after the manual migration choice is made the PastebinURL is reset meaning it won't trigger subsequently.

File: launcher/net/PasteUpload.cpp
Comment (Scrumplex) [reply launcher/net/PasteUpload.cpp:131]: Oh. I just read up again. Was looking at the wrong documentation.

File: launcher/Application.cpp
Comment (Scrumplex) [reply launcher/Application.cpp:928]: Brain lag. Makes sense

File: launcher/net/PasteUpload.cpp
Comment (LennyMcLennington) [reply launcher/net/PasteUpload.cpp:131]: > What Qt version is this targeting? Most of the codebase uses the `SIGNAL` and `SLOTG` macros

Pointer to member function is the modern way of doing it and has some advantages over the signal and slot macros. It's supported by the Qt versions we support.

File: launcher/Application.cpp
Comment (LennyMcLennington) [reply launcher/Application.cpp]: I agree.

Comment (LennyMcLennington) [reply launcher/Application.cpp]: I'll change it to this when I go on my laptop

Comment (LennyMcLennington) [reply launcher/Application.cpp]: Actually I can just commit it directly from GitHub

**Review by Scrumplex — APPROVED** [review]

**Comment by DioEgizio** [comment]
this is conflicting with develop

**Comment by LennyMcLennington** [comment]
> this is conflicting with develop

fixed

**Comment by LennyMcLennington** [comment]
This is what it looks like when they launch PolyMC after updating if this code gets merged as it is currently
image

Just so you know how it looks, if you want to tell me how much it sucks and suggest a better layout.

**Comment by DioEgizio** [comment]
looks good to me, but make that window a lot smaller

**Comment by DioEgizio** [comment]
also it'd be cool if you allowed changing the paste service directly in that screen

**Comment by LennyMcLennington** [comment]
> also it'd be cool if you allowed changing the paste service directly in that screen

I'll have a look at this in a little bit and see what I can do.

**Comment by LennyMcLennington** [comment]
> > also it'd be cool if you allowed changing the paste service directly in that screen
> 
> I'll have a look at this in a little bit and see what I can do.

It's possible but it's messy and imperfect and doesn't seem worth it for something that people will see 1 time.

**Comment by DioEgizio** [comment]
Alright, the important thing is that this gets merged today

**Review by d-513 — APPROVED** [review]

**Review by timoreo22 — APPROVED** [review]

**Pull request merged**
